src/bucket/BucketManagerImpl.cpp

```diff
@@ -395,6 +397,13 @@ BucketManagerImpl::adoptFileAsBucket(std::string const& filename,
         {
             auto timer = LogSlowExecution("Delete redundant bucket");
             std::remove(filename.c_str());
+
+            // race condition: two buckets + indexes were produced in parallel
```

Comment:
mm, not quite following the comment here - what's the race condition? (I might be blanking, but probably good to elaborate in the comment anyway)

Reply:
`adoptFileAsBucket` is supposed to be thread safe and called by both the main and worker threads. Sometimes, the same bucket exists in different levels of the BucketList. On startup, these identical buckets can be produced in parallel since they are on different levels and both call `adoptFileAsBucket`. In practice, there shouldn't be a race for the indexes since we check for this case in `IndexBucketsWork` and only index identical buckets once, but we keep `maybeSetIndex` here for safety.

Even though this shouldn't happen on startup, there are valid reasons to call `adoptFileAsBucket` on a bucket that already has an index after startup, so we can't just assert that the bucket is not indexed. For example, if a snap bucket merges with an empty bucket during normal operation after startup, the index will exist twice: once for the bucket currently part of the BucketList, and another for the result of the merge between that bucket and the empty level.

While we may not strictly need this race condition check since we try very hard not to double index buckets, it's there for defensive purposes.

Reply:
thanks for explaining, I seem to keep forgetting that we can produce identical valid buckets on different levels 😅